### PR TITLE
Use cookie parameters as set in the session

### PR DESCRIPTION
**Pull request by scottlucas**
Hi Marcel,

I've got another fix, this one is for regeneration cookies with non-default parameters such as long expiry, path, secure or HTTP only flags. Without this patch I end up with two distinct cookies in my browser when I use a non-existent session ID and browsers only replace cookies with matching parameters.

Thanks

Scott

### CODE REVIEW

**Comment by 1ma** [comment]
Hmm ok. Can you think of a way to test this edge case with Guzzle? Ideally I'd like a new test case that fails without the patch, like in your other PR.

**Comment by scottlucas** [comment]
Will do, TDD is good for the soul.

I'm working a bit blind without a working Docker instance but it is a good learning experience!

**Comment by 1ma** [comment]
Oh, does the latest version of `1maa/php-fpm:7.0` still fail on your box?

Edit: sorry, actually both have been updated, that is `1maa/php-fpm:7.0` and `1maa/php-cli:7.0`

**Comment by scottlucas** [comment]
I'm going to give it another try now. I have finished the commit with a few more changes.

These are images you have made yourself, with all the necessary dependencies right?

**Comment by scottlucas** [comment]
It is working now but not with "composer test".
I had to manually run a slightly different phpunit command as a script rather than through PHP:

`docker exec -t tests_testrunner_1 sh -c 'vendor/bin/phpunit'`

Thanks

**Comment by 1ma** [comment]
Correct. These are Alpine images that install PHP and its extensions as packages from their [official repository](https://pkgs.alpinelinux.org/packages?name=php7-*&branch=&repo=&arch=&maintainer=). Today I'd just use Docker's own PHP images, but their Alpine flavour did not exist when I made my own, and now I'm more used to these.

**Comment by 1ma** [comment]
Huh, that's weird. Feel free to remove "php" in your PR, on my box (Ubuntu 16.04 fwiw) it seems to work either way (as expected I'd say).

**Comment by scottlucas** [comment]
Very strange, maybe it is something to do with the version of PHPUnit as my vendor/bin/phpunit doesn't have any PHP in it. I am using Windows but it is in a Docker container anyway.

I've updated it in case anybody else wishes to do the same. I hope that phpredis will support this natively soon but it has taken long enough by the looks of it so wouldn't count on it.

**Comment by 1ma** [comment]
Dang, so that's the bug that marcing was talking about [here](https://github.com/1ma/RedisSessionHandler/issues/3#issuecomment-281938847), but I was using `ini_set` instead of `session_set_cookie_params` and was not able to reproduce it. Apparently when you set these parameters in the php.ini file `setcookie` [uses them implicitly](https://camo.githubusercontent.com/f52872aa3fd9a7fba767c01e85fe5dc589eeee54/687474703a2f2f692e696d6775722e636f6d2f504c6b416f57372e706e67), but not when you use `session_set_cookie_params`. Who knew!

File: tests/e2e/BasicTest.php [hunk 1]
```diff
@@ -162,4 +186,26 @@ protected function assertCreatedNewSession(Response $response)
...
+     * @param Response $response
+     */
+    protected function assertCustomCookieParams(Response $response)
+    {
+        $this->assertTrue($response->hasHeader('Set-Cookie'));
+        $this->assertStringStartsWith('PHPSESSID=', $response->getHeaderLine('Set-Cookie'));
+        $this->assertRegexp('/expires=/i', $response->getHeaderLine('Set-Cookie'));
+        $this->assertRegexp('/path=/i', $response->getHeaderLine('Set-Cookie'));
+        $this->assertRegexp('/secure/i', $response->getHeaderLine('Set-Cookie'));
+        $this->assertRegexp('/HttpOnly/i', $response->getHeaderLine('Set-Cookie'));
```
Comment (1ma) [inline tests/e2e/BasicTest.php]: Just one little thing: PHPStorm is whining that the actual method's name is `assertRegExp`, with a capital E.

File: README.md [hunk 2]
```diff
@@ -159,6 +159,8 @@ RedisSessionHandler solves this problem with a "lock" entry for every session th
 [Session fixation](https://www.owasp.org/index.php/Session_fixation) is the ability to choose your own session ID as an HTTP client. When clients are allowed to choose their
 session IDs, a malicious attacker might be able to trick other clients into using an ID already known to him, then wait for them log in and hijack their session.
 
+**To ensure cookies are replaced, any custom cookie parameters should be registered with session_set_cookie_params() as browsers will only update cookies with matching parameters.**
```
Comment (1ma) [inline README.md]: Dja think it's necessary to state anything? As far as I understand, now everything works as expected either if you set the custom session cookie params in the php.ini file or with `session_set_cookie_params` (just tested, it really works either way).

**Review by 1ma — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (scottlucas) [reply README.md]: If it works however the parameters are specified, then it is not necessary. I believe it will now work as parameters will always come from session_get_cookie_params() regardless. Your screenshot shows that session_start respects session.* parameters but setcookie only does this when specified explicitly. I am wondering whether there is a function we should actually be using such as session_regenerate_id() as then we are not coupling the session handler to usage of cookies. It remains to check whether this sends a new cookie but I will try it now.

I am trying to see if we have covered all the comments in the issue you referenced. I think the plugin is quite stable for most standard use cases now but testing may be required if you want to do something unusual (like run code with no time limit!). There are some knowledgeable comments but no pull requests!

**Comment by scottlucas** [comment]
I have tried to make the handler not reliant on the user using cookies but there are no tests.
I am not sure how this could be tested, for example if they are using URL based sessions or some other way of storing the identifier.

File: src/RedisSessionHandler.php [hunk 3]
```diff
@@ -137,9 +137,7 @@ public function create_sid()
     public function read($session_id)
     {
         if ($this->mustRegenerate($session_id)) {
-            session_id($session_id = $this->create_sid());
-
-            setcookie($this->cookieName, $session_id);
+            session_regenerate_id(true);
```
Comment (1ma) [inline src/RedisSessionHandler.php]: I used to call `session_regenerate_id` in the beginning, but it won't work on 7.1, see https://github.com/1ma/RedisSessionHandler/issues/3#issuecomment-281320830

Please revert to your original proposal :P

**Comment by 1ma** [comment]
It's OK, I don't want to support URL based sessions. These should not even exist in the first place (e.g. putting session IDs in a query param makes them easily loggable/cacheable along the software stack).

**Comment by scottlucas** [comment]
I understand they are insecure but if the user has their configuration set to enable them, we shouldn't break them. There aren't actually any significant changes to the code, just don't send a cookie if they are disabled.

**Comment by 1ma** [comment]
Yes but it's a terrible choice that I don't want to give the end user. For instance, the handler already ignores the `session.use_strict_mode` directive and always acts as if it were enabled (which usually isn't the case, by default it's `0`) because disabling strict mode does not make any sense.

I'm reluctant to support URL based sessions unless someone comes here and asks for it because they want to use the handler in their 20yo project.

**Comment by scottlucas** [comment]
I've reverted. It seemed impossible to get the URL based session fixation tests working without using session_regenerate_id() and I do not have the time to workout why. I've kept them in a branch in my repo anyway incase the question does get asked.

Perhaps there is a case for noting that certain configuration options behave differently than expected such as `session.use_strict_mode` and `session.use_cookies` but these are edge cases.

**Review by 1ma — APPROVED** [review]

**Comment by 1ma** [comment]
Thanks again :+1: Before I tag a new release, do you have anything else in store? :stuck_out_tongue:

**Comment by scottlucas** [comment]
Thanks for the code, I needed a working Redis handler with locks and learnt a bit along the way.
At the moment I'm all out of ideas, maybe before I can think of more phpredis will actually merge their patch!

**Comment by 1ma** [comment]
Hi @scottlucas , sorry to bother you. I've just expanded the integration test suite so that it tests all supported major versions of PHP (5.6, 7.0 and 7.1), but to do so I had to tweak the `composer test` script a little (see https://github.com/1ma/RedisSessionHandler/commit/49454676edc5b953fa54c657aa5854713d612a6f).

Would you be willing to check if this syntax still works in your local box?

**Comment by scottlucas** [comment]
No problem, the tests look good and everything is passing when I run the docker commands manually but for whatever reason my setup doesn't like the quotes in the test lines.

```
> docker exec -t tests_runner_1 sh -c 'TARGET=php56 vendor/bin/phpunit'
vendor/bin/phpunit': line 1: syntax error: unterminated quoted string
Script docker exec -t tests_runner_1 sh -c 'TARGET=php56 vendor/bin/phpunit' handling the test event returned with error code 2
```

Replacing with double quotes works but even the error message doesn't make sense as the quotes do terminate..

`"docker exec -t tests_runner_1 sh -c \"TARGET=php56 vendor/bin/phpunit\"",`

**Comment by 1ma** [comment]
Heh I knew it was worth asking, thank you. I've changed it, once again bash don't seem to care.

**Comment by 1ma** [comment]
Alright done, v0.9.5 released.
